### PR TITLE
Preserve selection when folding and unfolding blocks

### LINKED ISSUE #81: Preserve selection when folding and unfolding
Folding and unfolding code currently resets the cursor position and selection to none.

### PR DESCRIPTION
**Pull request by alexeyinkin**
- Resolves #81

### CODE REVIEW

File: lib/src/hidden_ranges/hidden_ranges.dart [hunk 1]
```diff
@@ -160,6 +160,67 @@ class HiddenRanges {
...
+    if (position >= ranges.last.end) {
+      return position - hiddenCharactersBeforeRanges.last;
+    }
+
+    int lowerRange = 0;
+    int upperRange = ranges.length - 1;
+
+    // Linear interpolation search.
+    while (upperRange > lowerRange) {
+      // Full characters at the lower ange start and at the upper range end.
```
Comment (Malarg) [inline lib/src/hidden_ranges/hidden_ranges.dart:181]: typo "range"

File: lib/src/hidden_ranges/hidden_ranges.dart [hunk 2]
```diff
@@ -160,6 +160,67 @@ class HiddenRanges {
...
+        return position - hiddenCharactersBeforeRanges[lowerRange];
+      }
+      if (rangeIndex > upperRange) {
+        return position - hiddenCharactersBeforeRanges[upperRange + 1];
+      }
+
+      final range = ranges[rangeIndex];
+
+      switch (range.compareToPosition(position)) {
+        case 0:
```
Comment (Malarg) [inline lib/src/hidden_ranges/hidden_ranges.dart:201]: MAy be will be more readable, if numbers will be changed on "equals", "upper", and "lower"

File: test/src/hidden_ranges/hidden_ranges_position_test.dart [hunk 3]
```diff
@@ -4,28 +4,120 @@ import 'package:flutter_code_editor/src/hidden_ranges/hidden_range.dart';
...
 void main() {
+  group('HiddenRanges. cutPosition', () {
+    test('No ranges - No changes', () {
+      expect(
+        HiddenRanges.empty.cutPosition(7),
+        7,
+      );
+    });
+
+    test('Cut with visible beginning and end', () {
```
Comment (Malarg) [inline test/src/hidden_ranges/hidden_ranges_position_test.dart:20]: This test does not cover case, where text contains hidden ranges at the start and at the end.

File: test/src/hidden_ranges/hidden_ranges_position_test.dart [hunk 4]
```diff
@@ -4,28 +4,120 @@ import 'package:flutter_code_editor/src/hidden_ranges/hidden_range.dart';
...
+        6: 5,
+        7: 5,
+        8: 5,
+        9: 5,
+        10: 5,
+        11: 6,
+      };
+      int i = 0;
+
+      for (final example in examples.entries) {
```
Comment (Malarg) [inline test/src/hidden_ranges/hidden_ranges_position_test.dart:104]: this loop repeat the same at 68 line. Suggest to extract

File: test/src/hidden_ranges/hidden_ranges_selection_test.dart [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        );
+        expect(
+          HiddenRanges.empty.recoverSelection(selection.reversed),
+          selection.reversed,
+        );
+      }
+    });
+
+    group('Collapsed', () {
+      test('before ranges', () {
```
Comment (Malarg) [inline test/src/hidden_ranges/hidden_ranges_selection_test.dart:67]: This 4 tests almost the same. Suggest to extract to a function and pass numbers as parameters

File: test/src/hidden_ranges/hidden_ranges_selection_test.dart [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        ),
+        TextSelection(
+          baseOffset: 5,
+          extentOffset: 10,
+          affinity: TextAffinity.upstream,
+          isDirectional: false,
+        ),
+      ];
+
+      for (final selection in selections) {
```
Comment (Malarg) [inline test/src/hidden_ranges/hidden_ranges_selection_test.dart:45]: May be this function helps to make this loop more clear
```
  void _expectTheSame(TextSelection selection) {
    expect(
      HiddenRanges.empty.cutSelection(selection),
      selection,
    );
    expect(
      HiddenRanges.empty.recoverSelection(selection),
      selection,
    );
  }
```

**Review by Malarg — CHANGES_REQUESTED** [review]

File: test/src/hidden_ranges/hidden_ranges_selection_test.dart (hunk 6 above)
Comment (alexeyinkin) [reply test/src/hidden_ranges/hidden_ranges_selection_test.dart:45]: This hides `expect` calls which is a bad pattern. They should be clearly visible. In 6 months when something breaks you will want to see `expect` on the top level.

File: test/src/hidden_ranges/hidden_ranges_position_test.dart (hunk 4 above)
Comment (alexeyinkin) [reply test/src/hidden_ranges/hidden_ranges_position_test.dart]: `+`

File: lib/src/hidden_ranges/hidden_ranges.dart [hunk 7]
```diff
@@ -160,6 +160,67 @@ class HiddenRanges {
     );
   }
 
+  /// Translates the [position] in the full text to the position in
+  /// the visible text.
+  ///
+  /// If [position] is hidden, returns the last visible position before it.
+  int cutPosition(int position) {
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/hidden_ranges/hidden_ranges.dart:167]: This is longer. `cut...` and `recover...` fit well.

File: lib/src/hidden_ranges/hidden_ranges.dart (hunk 2 above)
Comment (alexeyinkin) [reply lib/src/hidden_ranges/hidden_ranges.dart:201]: `-1`, `0`, and `1` are common for `sign`-like methods.

File: test/src/hidden_ranges/hidden_ranges_selection_test.dart (hunk 5 above)
Comment (alexeyinkin) [reply test/src/hidden_ranges/hidden_ranges_selection_test.dart]: Merged with the test below.

**Pull request merged**